pkg/pillar/kubeapi/nodedrain.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//go:build kubevirt
+
+package kubeapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/pubsub"
+)
+
+// RequestNodeDrain generates the NodeDrainRequest object and publishes it
+func RequestNodeDrain(pubNodeDrainRequest pubsub.Publication, requester DrainRequester, context string) error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("RequestNodeDrain: can't get hostname %v", err)
+	}
+	drainReq := NodeDrainRequest{
+		Hostname:    hostname,
+		RequestedAt: time.Now(),
+		RequestedBy: requester,
+		Context:     context,
+	}
+	err = pubNodeDrainRequest.Publish("global", drainReq)
+	if err != nil {
+		return fmt.Errorf("RequestNodeDrain: error publishing drain request: %v", err)
+	}
+	return nil
+}
+
+// GetDrainStatusOverride : an alternate way to set drain status for debug
+func GetDrainStatusOverride(log *base.LogObject) *NodeDrainStatus {
+	// An alternate path to force a drain status in the event of a drain issue.
+	forceNodeDrainPath := "/persist/kube-status/force-NodeDrainStatus-global.json"
+	if _, err := os.Stat(forceNodeDrainPath); err != nil {
+		return nil
+	}
+	b, err := os.ReadFile(forceNodeDrainPath)
+	if err != nil {
+		log.Warnf("Unable to read %s:%v", forceNodeDrainPath, err)
+		return nil
+	}
+	cfg := NodeDrainStatus{}
+	err = json.Unmarshal(b, &cfg)
+	if err != nil {
+		log.Warnf("Unable to Unmarshal %s to NodeDrainStatus: %v", forceNodeDrainPath, err)
```

Comment:
This is not sufficiently robust in the face of power outages.
Even with WriteRename() the file can be zero length (the meta-data was written/logged but not the data).
In such a case do you want to look for an older backup version of the file, or do you want to delete the empty file and pretend it was never written?

Reply:
This path is only reading a file which is written out by a user in dev/debug and filling in the NodeDrainStatus.  Although I do see that the zedkube startup path needs to delete this file to handle the case of a power loss before the COMPLETE state is met.